src/code/z_bgcheck.c

```diff
@@ -1,461 +1,4293 @@
+s32 func_8003C55C(CollisionContext* colCtx, Vec3f* pos) {
+    if (pos->x < (colCtx->minBounds.x - 50.0f) || (colCtx->maxBounds.x + 50.0f) < pos->x ||
+        pos->y < (colCtx->minBounds.y - 50.0f) || (colCtx->maxBounds.y + 50.0f) < pos->y ||
+        pos->z < (colCtx->minBounds.z - 50.0f) || (colCtx->maxBounds.z + 50.0f) < pos->z) {
+        return false;
+    }
+    return true;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041E4C.s")
+// Raycast Toward Floor
+// returns the yIntersect of the nearest poly found directly below `pos`, or BGCHECK_Y_MIN if no floor detected
+// returns the poly found in `outPoly`, and the bgId of the entity in `outBgId`
+f32 func_8003C614(GlobalContext* globalCtx, CollisionContext* colCtx, u16 xpFlags, CollisionPoly** outPoly,
+                  s32* outBgId, Vec3f* pos, Actor* actor, u32 arg7, f32 chkDist) {
+
+    f32 yIntersectDyna;
+    s32* temp_a0;
+    Lookup* lookupTbl;
+    Vec3f checkPos;
+    Lookup* lookup;
+    DynaRaycast dynaRaycast;
+    f32 yIntersect;
+
+    *outBgId = BGCHECK_SCENE;
+    *outPoly = NULL;
+    lookupTbl = colCtx->lookupTbl;
+    yIntersect = BGCHECK_Y_MIN;
+    checkPos = *pos;
+
+    while (true) {
+        if (checkPos.y < colCtx->minBounds.y) {
+            break;
+        }
+        if (BgCheck_PosErrorCheck(&checkPos, "../z_bgcheck.c", 4410)) {
+            if (actor != NULL) {
+                osSyncPrintf("こいつ,pself_actor->name %d\n", actor->id);
+            }
+        }
+        lookup = func_8003AD00(colCtx, lookupTbl, &checkPos);
+        if (lookup == NULL) {
+            checkPos.y -= colCtx->unitSize.y;
+            continue;
+        }
+        yIntersect = func_8003992C(lookup, colCtx, xpFlags, outPoly, pos, arg7, chkDist, BGCHECK_Y_MIN);
+        if (BGCHECK_Y_MIN < yIntersect) {
+            break;
+        }
+        checkPos.y -= colCtx->unitSize.y;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041EC8.s")
+    dynaRaycast.colCtx = colCtx;
+    dynaRaycast.xpFlags = xpFlags;
+    dynaRaycast.yIntersect = yIntersect;
+    dynaRaycast.pos = pos;
+    dynaRaycast.actor = actor;
+    dynaRaycast.unk_20 = arg7;
+    dynaRaycast.chkDist = chkDist;
+    dynaRaycast.globalCtx = globalCtx;
+    dynaRaycast.resultPoly = outPoly;
+    dynaRaycast.bgId = outBgId;
+
+    yIntersectDyna = func_8003FDDC(&dynaRaycast);
+
+    if (yIntersect < yIntersectDyna) {
+        yIntersect = yIntersectDyna;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_bgcheck/func_80041EEC.s")
+    if (yIntersect != BGCHECK_Y_MIN) {
+        if (func_80041EC8(colCtx, *outPoly, *outBgId) != 0) {
+            yIntersect -= 1.0f;
+        }
+    }
+    return yIntersect;
+}
 
-u32 func_80041F10(u32 a0, u32 a1, u32 a2) {
-    u32 uVar1;
+// Public raycast toward floor
+// returns yIntersect of the poly found, or BGCHECK_Y_MIN if no poly detected
+f32 func_8003C834(CollisionContext* colCtx, CollisionPoly** outPoly, Vec3f* pos) {
+    s32 bgId;
```

Comment:
naming the raycast functions would be cool if we can do that/understand the differences